src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/ControlPlaneDataStore.cs

```diff
@@ -0,0 +1,142 @@
+            return resultIdentityProvider.ToIdentityProvider();
+        }
+
+        internal IDocumentQuery<T> CreateDocumentQuery<T>(
+            SqlQuerySpec sqlQuerySpec,
+            FeedOptions feedOptions = null)
+        {
+            EnsureArg.IsNotNull(sqlQuerySpec, nameof(sqlQuerySpec));
+
+            var context = new CosmosQueryContext(_collectionUri, sqlQuerySpec, feedOptions);
+
+            return _cosmosDocumentQueryFactory.Create<T>(_documentClient.Value, context);
+        }
+
+        private async Task<T> GetSystemDocumentByIdAsync<T>(string id, string partitionKey, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(id, nameof(id));
+            EnsureArg.IsNotNull(partitionKey, nameof(partitionKey));
+            var documentQuery = new SqlQuerySpec(
+                "SELECT * FROM root r WHERE r.id = @id",
+                new SqlParameterCollection(new[] { new SqlParameter("@id", id) }));
+            var feedOptions = new FeedOptions
+            {
+                PartitionKey = new PartitionKey(partitionKey),
+            };
+            IDocumentQuery<T> cosmosDocumentQuery =
+                CreateDocumentQuery<T>(documentQuery, feedOptions);
+            using (cosmosDocumentQuery)
+            {
+                var response = await cosmosDocumentQuery.ExecuteNextAsync(cancellationToken);
+                var result = response.SingleOrDefault();
+                if (result == null)
+                {
+                    _logger.LogError($"{typeof(T)} with id {id} was not found in Cosmos DB.");
+                }
+
+                return result;
+            }
+        }
+
+        private async Task<T> UpsertSystemObjectAsync<T>(T systemObject, string partitionKey, CancellationToken cancellationToken)
+            where T : class
+        {
+            EnsureArg.IsNotNull(systemObject, nameof(systemObject));
+            var eTagAccessCondition = new AccessCondition();
+
+            var requestOptions = new RequestOptions
+            {
+                PartitionKey = new PartitionKey(partitionKey),
+                AccessCondition = eTagAccessCondition,
+            };
+            try
+            {
+                var response = await _documentClient.Value.UpsertDocumentAsync(
+                    _collectionUri,
+                    systemObject,
+                    requestOptions,
+                    true,
+                    cancellationToken);
+                _logger.LogInformation("Request charge: {RequestCharge}, latency: {RequestLatency}", response.RequestCharge, response.RequestLatency);
+                return (dynamic)response.Resource;
+            }
+            catch (DocumentClientException dce)
+            {
+                _logger.LogError(dce, "Unhandled Document Client Exception");
+                throw;
```

Comment:
UpsertIdentityProviderAsync is called by RbacService which sits in the Core and since it doesn't know anything about the underlying persistent layer, it doesn't understand DocumentClientException. Should we normalize it into some exception it understands so it can be handled later? Just some food for thought

Reply:
I added a check for the 429 exception, but we should probably come up with something to wrap the persistence specific errors into. I'll open an issue for that portion.

Reply:
We can use the code generator for that to create an implementation of IDocumentClient that checks the common exception like this and also log the RU usage so that it can be used both by CP and Fhir. Let's open the issue for that and we can work on it separately.